rust/lance-datagen/src/generator.rs

```diff
@@ -0,0 +1,1126 @@
+    fn from(n: u64) -> Self {
+        Self(n)
+    }
+}
+
+impl From<u64> for ByteCount {
+    fn from(n: u64) -> Self {
+        Self(n)
+    }
+}
+
+impl From<u32> for Dimension {
+    fn from(n: u32) -> Self {
+        Self(n)
+    }
+}
+
+/// A trait for anything that can generate arrays of data
+pub trait ArrayGenerator {
+    /// Generate an array of the given length
+    ///
+    /// # Arguments
+    ///
+    /// * `length` - The number of elements to generate
+    /// * `rng` - The random number generator to use
+    ///
+    /// # Returns
+    ///
+    /// An array of the given length
+    ///
+    /// Note: Not every generator needs an rng.  However, it is passed here because many do and this
+    /// lets us manage RNGs at the batch level instead of the array level.
+    fn generate(
+        &mut self,
+        length: RowCount,
+        rng: &mut rand_xoshiro::Xoshiro256PlusPlus,
+    ) -> Result<Arc<dyn arrow_array::Array>, ArrowError>;
+    /// Get the data type of the array that this generator produces
+    ///
+    /// # Returns
+    ///
+    /// The data type of the array that this generator produces
+    fn data_type(&self) -> &DataType;
+    /// Get the size of each element in bytes
+    ///
+    /// # Returns
+    ///
+    /// The size of each element in bytes.  Will be None if the size varies by element.
+    fn element_size_bytes(&self) -> Option<ByteCount>;
+}
+
+pub struct NullGenerator {
+    generator: Box<dyn ArrayGenerator>,
+    null_probability: f64,
+}
+
+impl ArrayGenerator for NullGenerator {
+    fn generate(
+        &mut self,
+        length: RowCount,
+        rng: &mut rand_xoshiro::Xoshiro256PlusPlus,
+    ) -> Result<Arc<dyn arrow_array::Array>, ArrowError> {
+        let array = self.generator.generate(length, rng)?;
+        let data = array.to_data();
+
+        if self.null_probability < 0.0 || self.null_probability > 1.0 {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "null_probability must be between 0 and 1, got {}",
+                self.null_probability
+            )));
+        }
+
+        let (null_count, new_validity) = if self.null_probability == 0.0 {
+            if data.null_count() == 0 {
+                return Ok(array);
+            } else {
+                (0_usize, None)
+            }
+        } else if self.null_probability == 1.0 {
+            if data.null_count() == data.len() {
+                return Ok(array);
+            } else {
+                let all_nulls = BooleanBuffer::new_unset(array.len());
+                (array.len(), Some(all_nulls.into_inner()))
+            }
+        } else {
+            let array_len = array.len();
+            let num_validity_bytes = (array_len + 7) / 8;
+            let mut null_count = 0;
+            // Sampling the RNG once per bit is kind of slow so we do this to sample once
+            // per byte.  We only get 8 bits of RNG resolution but that should be good enough.
```

Comment:
Can we add a simple statistical test to make sure we're getting a decent range and uniformity? (e.g. if we sample 1M values and aggregate them into 5 equisized brackets over the entire range we get at least 150k in each).